### PR TITLE
Fix simulcast video quality issue due to layer selection

### PR DESCRIPTION
**Pull request by harryz2000**

### CODE REVIEW

File: test/unit/TestSimulcastMediaFrameListener.cpp [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+{
+public:
+	static constexpr uint8_t	LOW = 0x1;
+	static constexpr uint8_t	MID = 0x2;
+	static constexpr uint8_t	HIGH = 0x4;
+	static constexpr uint8_t	INTRA = 0x8;
+	static constexpr uint8_t	ALL_NON_INTRA = LOW | MID | HIGH;
+	static constexpr uint8_t	ALL_INTRA = LOW | MID | HIGH | INTRA;
+
+	class Excutor
```
Comment (murillo128) [inline test/unit/TestSimulcastMediaFrameListener.cpp:114]: Remove executor, use PushAdvance directly instead of <<

Comment (harryz2000) [reply test/unit/TestSimulcastMediaFrameListener.cpp]: Removed.

File: test/unit/TestSimulcastMediaFrameListener.cpp [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+		frame->SetLength(input.width*input.width);
+		frame->SetClockRate(ClockRate);	
+
+		listener.onMediaFrame(frame->GetSSRC(), *frame);
+	}
+
+	void CheckResetForwardedFrames(const std::vector<std::tuple<uint32_t, uint64_t, uint64_t>>& expected)
+	{
+		auto forwarded = std::move(mediaFrameListener->forwardedFrames);
+		size_t count = std::min(expected.size(), forwarded.size());
```
[the remark this answers is not in the document]
Comment (harryz2000) [reply test/unit/TestSimulcastMediaFrameListener.cpp]: Done.

File: test/unit/TestSimulcastMediaFrameListener.cpp [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+		for (size_t i = 0; i < count; i++)
+		{
+#ifdef CHECK_WIDTH_ONLY
+			ASSERT_EQ(std::get<0>(expected[i]), std::get<0>(forwarded[i])) << "index:" << i << "\n" 
+					<< "expected: width:" << std::get<0>(expected[i]) << ",\ttimestamp:" << std::get<1>(expected[i]) << ",\ttime:" << std::get<2>(expected[i]) << "\n"
+					<< "  actual: width:" << std::get<0>(forwarded[i]) << ",\ttimestamp:" << std::get<1>(forwarded[i]) << ",\ttime:" << std::get<2>(forwarded[i]);
+#else
+			ASSERT_EQ(expected, forwarded);
```
[the remark this answers is not in the document]
Comment (harryz2000) [reply test/unit/TestSimulcastMediaFrameListener.cpp]: This line shouldn't have been inside the loop. Fixed.

File: src/SimulcastMediaFrameListener.cpp [hunk 4]
```diff
@@ -33,221 +35,156 @@ void SimulcastMediaFrameListener::RemoveMediaListener(const MediaFrame::Listener
...
 
-		// If we have enough timed out waiting
-		if (frameTime > selectionTime + FrameWaitingTime)
-			//Don't wait for other layers and force layer selection
-			Select();
-
-		//Done
-		return;
+	// forward the front frame if the queue is full
+	if (queue.size() > MaxQueueSize)
```
[the remark this answers is not in the document]
Comment (harryz2000) [reply src/SimulcastMediaFrameListener.cpp]: Now calculate the layers basing on layers.

File: src/SimulcastMediaFrameListener.cpp [hunk 5]
```diff
@@ -33,221 +47,217 @@ void SimulcastMediaFrameListener::RemoveMediaListener(const MediaFrame::Listener
...
 
-		//Done
-		return;
+		assert(!timestampLayers.empty());
+		// forward the front frame if the queue is full or the time indicates the front
+		// is the earliest possible eligible frame
+		if (queue.size() > maxQueueSize || timestampLayers.front().second.size() == numLayers)
+		{
+			auto f = std::move(queue.front());
+			f->SetSSRC(0);
```
Comment (murillo128) [inline src/SimulcastMediaFrameListener.cpp:243]: why 0?

Comment (harryz2000) [reply src/SimulcastMediaFrameListener.cpp:243]: Just give it a unified ssrc. It seems the recording is not using the ssrc. Or any value do you think is better?

Comment (harryz2000) [reply src/SimulcastMediaFrameListener.cpp]: Double checked it looks the ssrc passed from constructor should be a better option. Changed.

File: src/SimulcastMediaFrameListener.cpp [hunk 6]
```diff
@@ -33,221 +47,195 @@ void SimulcastMediaFrameListener::RemoveMediaListener(const MediaFrame::Listener
...
+		//Uh?
+		return;
 
-	DWORD prev = forwarded;
-	VideoFrame* selected = nullptr;
-	DWORD currentDimensions = 0;
-	DWORD currentSize = 0;
+	//Get cloned video frame
+	auto cloned = std::unique_ptr<VideoFrame>((VideoFrame*)frame.Clone());
+	cloned->SetSSRC(ssrc);
```
Comment (murillo128) [inline src/SimulcastMediaFrameListener.cpp:88]: you are overriding the ssrc here

File: src/SimulcastMediaFrameListener.cpp [hunk 7]
```diff
@@ -33,221 +47,195 @@ void SimulcastMediaFrameListener::RemoveMediaListener(const MediaFrame::Listener
...
-	if (prev!=forwarded)
+	//Send it to all listeners
+	for (const auto& listener : listeners)
+		//Send cloned frame
+		listener->onMediaFrame(ssrc, frame);
+}
+
+void SimulcastMediaFrameListener::Push(std::unique_ptr<VideoFrame> frame)
+{
+	DWORD ssrc = frame->GetSSRC();
```
Comment (murillo128) [inline src/SimulcastMediaFrameListener.cpp:114]: this ssrc is overriden in the onMediaFrame

Comment (harryz2000) [reply src/SimulcastMediaFrameListener.cpp:114]: Renamed the member one to forwardSsrc.

File: src/SimulcastMediaFrameListener.cpp (hunk 6 above)
Comment (harryz2000) [reply src/SimulcastMediaFrameListener.cpp:88]: As discussed, this is a local ssrc. Renamed the member one to forwardSsrc.

File: src/SimulcastMediaFrameListener.cpp [hunk 8]
```diff
@@ -33,221 +47,195 @@ void SimulcastMediaFrameListener::RemoveMediaListener(const MediaFrame::Listener
...
-
-	//Discard out of order frames
-	if (ts < firstTimestamp)
+	while(!queue.empty())
 	{
-		//Error
-		Warning("SimulcastMediaFrameListener::ForwardFrame() | Discarding out of order frame [ts:%llu,first:%llu]\n", ts, firstTimestamp);
-		return;
+		auto f = std::move(queue.front());
+		queue.pop_front();
```
Comment (murillo128) [inline src/SimulcastMediaFrameListener.cpp:238]: Missing `f->SetSSRC(forwardSsrc);` or do it inside `ForwardFrame`

Comment (harryz2000) [reply src/SimulcastMediaFrameListener.cpp:238]: Moved to be inside ForwardFrame.

**Review by murillo128 — APPROVED** [review]

**Pull request merged**